bpf: optmize builtin functions before we fallback to them

borkmann opened the pull request:
See commit msg.

borkmann committed:
bpf: optimized memcpy/memzero with dw-wide copies

While checking some of the xlated code from XDP side, I recently
noticed __builtin_memcpy() patterns like:

  [...]
      14:    55 04 ee 00 00 00 00 00    if r4 != 0 goto +238 <LBB0_2>
      15:    71 32 2f 00 00 00 00 00    r2 = *(u8 *)(r3 + 47)
      16:    67 02 00 00 08 00 00 00    r2 <<= 8
      17:    71 31 2e 00 00 00 00 00    r1 = *(u8 *)(r3 + 46)
      18:    4f 12 00 00 00 00 00 00    r2 |= r1
      19:    7b 2a 78 ff 00 00 00 00    *(u64 *)(r10 - 136) = r2
      20:    71 32 37 00 00 00 00 00    r2 = *(u8 *)(r3 + 55)
      21:    67 02 00 00 08 00 00 00    r2 <<= 8
      22:    71 31 36 00 00 00 00 00    r1 = *(u8 *)(r3 + 54)
      23:    4f 12 00 00 00 00 00 00    r2 |= r1
      24:    7b 2a b0 ff 00 00 00 00    *(u64 *)(r10 - 80) = r2
  [...]

This is bad since we end up doing byte-wise copy of data. LLVM is
not aware of efficient unaligned access on x86-64 or arm64 etc so
it cannot make any assumptions on the access.

Implement optimized routines and make sure we don't blindly use
pain __builtin_*() directly in the code.

  [...]
      15:    79 31 38 00 00 00 00 00    r1 = *(u64 *)(r3 + 56)
      16:    7b 1a f8 ff 00 00 00 00    *(u64 *)(r10 - 8) = r1
      17:    79 31 30 00 00 00 00 00    r1 = *(u64 *)(r3 + 48)
      18:    7b 1a f0 ff 00 00 00 00    *(u64 *)(r10 - 16) = r1
      19:    79 31 28 00 00 00 00 00    r1 = *(u64 *)(r3 + 40)
      20:    7b 1a e8 ff 00 00 00 00    *(u64 *)(r10 - 24) = r1
      21:    79 31 20 00 00 00 00 00    r1 = *(u64 *)(r3 + 32)
      22:    7b 1a e0 ff 00 00 00 00    *(u64 *)(r10 - 32) = r1
      23:    79 31 18 00 00 00 00 00    r1 = *(u64 *)(r3 + 24)
      24:    7b 1a d8 ff 00 00 00 00    *(u64 *)(r10 - 40) = r1
  [...]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: no more direct use of __builtin_mem{cpy,set} in bpf

Convert all direct use of __builtin_mem{cpy,set}() over to the
regular mem{cpy,set}() function as we're going to have a custom
implementation of the latter two.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno commented on bpf/include/bpf/builtins.h:
```diff
@@ -10,12 +10,80 @@
+					    const __u64 len)
+{
+	switch (len) {
+#define __it_set(a, b, op) (*(__u##op *)__it(a, op)) = b
```
Don't we need to copy `b` several times since it's `__u8` and we want to write a `__u{64,32}`?

borkmann replied:
I haven't looked at the output of `__bpf_ua_memset()` yet, only the `__bpf_ua_memcpy()` so far, so I'm going to do that now. :-)

borkmann replied:
The output looks reasonable to me, for example, clang generates:
```
       1:	b7 04 00 00 00 00 00 00	r4 = 0
      [...]
      15:	7b 43 f8 ff 00 00 00 00	*(u64 *)(r3 - 8) = r4
      16:	7b 43 f0 ff 00 00 00 00	*(u64 *)(r3 - 16) = r4
      17:	7b 43 e8 ff 00 00 00 00	*(u64 *)(r3 - 24) = r4
      18:	7b 43 e0 ff 00 00 00 00	*(u64 *)(r3 - 32) = r4
      19:	7b 43 d8 ff 00 00 00 00	*(u64 *)(r3 - 40) = r4
      20:	7b 43 d0 ff 00 00 00 00	*(u64 *)(r3 - 48) = r4
      21:	7b 43 c8 ff 00 00 00 00	*(u64 *)(r3 - 56) = r4
      22:	7b 43 c0 ff 00 00 00 00	*(u64 *)(r3 - 64) = r4

```

pchaigno replied:
Because `r4 = 0`. If you pass a non-zero value, you'll end up writing only 8 bits of those 64; the others will be remain zero. Or am I missing something?

borkmann replied:
I'm not sure I understand your concern yet. We're doing a `*(u64 *)(r3 - 8) = ...`, why would we end up only assigning 1 byte?

qmonnet replied:
I talked about it with Paul and there are two points that we didn't manage to figure out.

1) As Paul mentioned, the constant copied is a `__u8` and it seems that would result in a value up to 255 being assigned to a 64-bit integer. With your example:
    ```
       1:	b7 04 00 00 00 00 00 00	r4 = 3    // changed here
      [...]
      15:	7b 43 f8 ff 00 00 00 00	*(u64 *)(r3 - 8) = r4
      16:	7b 43 f0 ff 00 00 00 00	*(u64 *)(r3 - 16) = r4
      17:	7b 43 e8 ff 00 00 00 00	*(u64 *)(r3 - 24) = r4
    ```
   It feels like we will end up with
    ```
    r3 == 00 00 00 00 00 00 00 03   00 00 00 00 00 00 00 03   00 00 00 00 00 00 00 03
    ````
    Instead of setting each byte:
    ```
    r3 == 03 03 03 03 03 03 03 03   03 03 03 03 03 03 03 03   03 03 03 03 03 03 03 03
    ````

2) We also struggle to understand the subtraction in `#define __it(x, op) (x -= sizeof(__u##op))`. It makes this new `memset()` start from the beginning of the destination and go backwards. Should it not go forwards with a `+` instead?

borkmann replied:
Re 1) Ahh, good point. I see now what you mean. Yep, I'll change the name to bzero() or so, since we don't need anything else other than zero'ing in our code.

Re 2) Yep, I just noticed here. Missing an initial ptr+=len, since we walk backwards.

pchaigno replied:
> since we don't need anything else other than zero'ing in our code.

Except for [`test_ipv6_addr_clear_suffix()`](https://github.com/cilium/cilium/blob/f96afb0821ca95a23429704604e8707c043ec94d/bpf/lib/ipv6_test.h#L9), the failing Travis CI test, but we probably don't really **need** it.

borkmann replied:
Fyi, still figuring out what we can do wrt stack alignment. Sadly the verifier forces BPF access to stack as aligned (even x86-64). Digging up a bit the LLVM code to see if there's a reasonable workaround.

tklauser reviewed (commented):
Small nit: s/optmize/optimize/ in commit message and PR title.

pchaigno commented on bpf/include/bpf/builtins.h:
```diff
@@ -6,17 +6,240 @@
+static __always_inline void __bpf_memzero(void *d, __u64 len)
+{
+#if __clang_major__ >= 10
+	const __u64 s = 0;
```
Couldn't we get rid of `s` since it's always zero and simplify the macro?

borkmann replied:
Ah yeah, code generation wise, it shouldn't make a difference though.

pchaigno commented on bpf/include/bpf/builtins.h:
```diff
@@ -6,17 +6,240 @@
+	case 38: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_32;
+	case 30: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_24;
+	case 22: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_16;
+	case 14: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_8;
```
Would be nice to have a bunch of unit tests with different sizes for this logic, but I'm guessing the `__non_bpf_context` guard will make that complicated...?

borkmann replied:
No, we could. I'll see to add them. The guard is only since the bpf include infra is also included from the map-migrate tool.

pchaigno commented on bpf/include/bpf/builtins.h:
```diff
@@ -6,17 +6,240 @@
+ * for <= 8 bytes and in case of > 8 bytes only when 8 byte is not
+ * the natural alignment.
+ */
+#define __align_stack_8		__aligned(8)
```
Would be nice to have some static test that we are using this macro everywhere we should., but I'm not sure I understand when we should use it. For all on-stack objects of size >8B? Only for those that get used in `memcpy`/`memset`? This commit seems to do neither, so I'm probably missing something.

borkmann replied:
On stack objects >8b that are passed in either memcpy/memset and have a natural alignment of < 8b, e.g. `__u8 foo[12]`. Turns out we don't have that many objects with such property, so it's not needed often.

pchaigno replied:
Would be nice to have a static check for that. Not sure Coccinelle is the best fit to do that (I don't think it can match on objects' sizes and alignments), but I can have a look. Worth-case scenario, we hardcode the known-problematic structures in the semantic patch.

borkmann replied:
My smpl skills are only very rusty, I tried coming up with something yesterday night, but got stuck here basically. I'm probably looking at it from the wrong angle in that the below might not be possible this way as in finding the declaration of the expression `T *x` we pass in below and adding the attribute to it. I got stuck at the mix of expression from the match and getting the identifier for m.x:
```
@m@
expression e1, e2, e3, e4;
type T;
T *x;
@@
(
        ctx_load_bytes(e1, e2, x, e3);
|
        ctx_store_bytes(e1, e2, x, e3, e4);
)

@@
attribute name __align_stack;
//expression m.x
identifier x;
type T;
@@
-       T x;
+       T x __align_stack;
```
If you find something I would be very curious! :-)

borkmann replied:
Agree that it might be easier finding the bad cases explicitly by name though. Also, for most cases in our code base things will just work due to natural object alignment, so it's really just a few cases luckily.

pchaigno reviewed (changes requested):
This is great! One nit and a couple questions below.

Did you get a chance to check the complexity and program size impact? It looks like it could have a bigger impact than other mitigations we've discussed, and on all kernels :tada:

borkmann committed:
bpf: move all unit tests to bpf/tests/

Place them into its own directory instead of directly into lib/
code. Adding more into lib/ would convolute it too much. Also
add a Wno-builtin-declaration-mismatch for gcc.

Tested via: # make unit-tests TESTPKGS=

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add bpf unit tests for builtin replacements

Add various new test cases for mem{cpy,set} adn run them as part
of the bpf unit tests. I've added barrier_data() from [0] to
prevent the compiler from any optimizations on the test data.

Tested via: # make unit-tests TESTPKGS=

  [0] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7829fb09a2b4268b30dd9bc782fa5ebee278b137

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

qmonnet reviewed (approved):
Very nice set! Thanks
